### ISSUE TITLE
C++ exceptions are not caught on macOS

### ISSUE THREAD

**Issue by vaivaswatha** [ocaml/ocaml#9026]
When interfacing with C++ code, exceptions that are supposed to be caught within the C++ code itself are not caught on macOS. 

```cpp
try {
  // throw something
} catch (...) {
  // control does not come here at all
}
```

I have attached a zip file with an independent + miniature test case to reproduce the problem. Please check the `Readme.md` of the zip file.

The same program on macOS produces `libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: This is not caught on macOS, why?`, but on Ubuntu produces `Exception successfully caught`.

[OCamlCppExn.zip](https://github.com/ocaml/ocaml/files/3703935/OCamlCppExn.zip)

**Comment by nojb** [comment]
You need to pass `-Wl,-keep_dwarf_unwind` to the linker, see https://gitlab.haskell.org/ghc/ghc/issues/11829.

**Comment by vaivaswatha** [comment]
Thank you @nojb . That helped.

I did come across that thread you pointed out when I Googled for this problem, however I couldn't find any documentation on `-keep_dwarf_unwind` (it isn't listed on the arguments list for `clang` or `ld`) and ended up assuming that the thread is outdated. I couldn't immediately figure out how to pass those options with `dune` and just left it at that :-(. Sorry about that.

I've added the sample code here https://github.com/vaivaswatha/OCamlCppExn in case someone finds this thread and wants to see the build config files.

My colleague has also filed an Issue with `dune` on how to enable providing this flag only on macOS (since it fails on GNU/Linux as an unknown flag). https://github.com/ocaml/dune/issues/2718

Thanks again. Closing this.
